src/phoenix/experimental/evals/long_context_evaluators.py

```diff
@@ -0,0 +1,58 @@
+from typing import List, Optional
+
+from phoenix.experimental.evals import PromptTemplate
+from phoenix.experimental.evals.models import BaseEvalModel
+
+
+class Refiner:
+    def __init__(
+        self,
+        model: BaseEvalModel,
+        initial_prompt_template: PromptTemplate,
+        refine_prompt_template: PromptTemplate,
+        reduce_prompt_template: Optional[PromptTemplate] = None,
+    ) -> None:
+        self._model = model
+        self._initial_prompt_template = initial_prompt_template
+        self._refine_prompt_template = refine_prompt_template
+        self._reduce_prompt_template = reduce_prompt_template
+
+    def evaluate(self, chunks: List[str]) -> str:
+        if len(chunks) < 2:
+            raise ValueError("Must provide two or more documents")
+        model = self._model
+        initial_prompt = self._initial_prompt_template.format({"chunk": chunks[0]})
+        accumulator = model(initial_prompt)
+        for chunk in chunks[1:]:
+            refine_prompt = self._refine_prompt_template.format(
+                {"accumulator": accumulator, "chunk": chunk}
+            )
+            accumulator = model(refine_prompt)
+        if not self._reduce_prompt_template:
+            return accumulator
+        reduce_prompt = self._reduce_prompt_template.format({"accumulator": accumulator})
+        return model(reduce_prompt)
```

Comment:
can you help me understand why a final prompting would be needed? Shouldn't the refine template do a good enough job?

Reply:
My thought here is that the accumulator contains not a single classification or rail (e.g., "accurate" or "not_accurate" for summaries), but a verbose evaluation and explanation of the summary relative to the context it has encountered thus far. This is not the approach taken by LlamaIndex, which seems to accumulate just "YES" or "NO" for evaluations. Intuitively, that does not make sense to me, since there isn't anything being accumulated in a binary value. If we want to reduce the verbose accumulated explanation down to a classification at the end, we need a final synthesis or reduce step at the end.